### PR TITLE
Improved the error message when using "@" in a decorated service

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17666 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by stloyd** [comment]
:+1:

**Comment by dunglas** [comment]
:+1: (I made this mistake recently).

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 1]
```diff
@@ -288,6 +288,10 @@ private function parseDefinition($id, $service, $file)
         }
 
         if (isset($service['decorates'])) {
+            if ('@' === $service['decorates']{0}) {
```
Comment (gharlan) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: Could `$service['decorates']` be empty here?

Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: and by the way, I don't think we use braces anywhere else to access chars in string (everywhere else we use square brackets instead IIRC)

**Comment by javiereguiluz** [comment]
I've implemented the changes asked by reviewers. This is now ready for the final review. Thanks!

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 2]
```diff
@@ -288,6 +288,10 @@ private function parseDefinition($id, $service, $file)
         }
 
         if (isset($service['decorates'])) {
+            if ('' !== $service['decorates'] && '@' === $service['decorates'][0]) {
+                throw new InvalidArgumentException(sprintf('The value of the "decorates" option for the "%s" service must be the id of the service without the "@" prefix (replace "%s" by "%s").', $id, $service['decorates'], substr($service['decorates'], 1)));
```
Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: [...] replace with [...]

**Comment by xabbuh** [comment]
I left a minor comment, but beside that :+1:

**Comment by Tobion** [comment]
AFAIK you can technically have a service id like `@foo`. The limitation is only in the PhpDumper (https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L1342) (which is arbitrary and could be fixed in the future).
So this change is not consistent with the rest and could break code that actually references a service beginning with an `@`.

**Comment by Tobion** [comment]
If I'm right, we need a different solution.

**Comment by stof** [comment]
@Tobion IMO, the right solution is to restrict the list of allowed chars in service ids to chars working everywhere (anything not working for the PhpDumper means it cannot work in a project wanting performance, as it would prevent dumping the container in the cache). This could start as a deprecation for now though

**Comment by Tobion** [comment]
Yes that is probably the easiest solution. Another one would be the remove this restriction in the PhpDumper by encoding special chars (hex for example).

And IMO the better solution this is issue here is to improve the real exception message in the ContainerBuilder: mention which service is decorating the non-existing one and maybe add suggestions for existing ones (like it is done for console commands).

**Comment by fabpot** [comment]
I'm :+1: to merge this one as is (even if technically, a service name can start with `@`, nobody will ever do that anyway). @javiereguiluz Can you create another PR or an issue to forbid usage of characters not supported by the dumper?

**Comment by javiereguiluz** [comment]
Yes, I'll take care of that in the new issue #17801.

**Comment by xabbuh** [comment]
But wouldn't it indeed be better to raise this exception only in the `DecoratorServicePass` if the `getDecoratedService()` method returns a non-existent service or alias id?

**Comment by fabpot** [comment]
@xabbuh I think it's better to throw the exception as early as possible; it is a YAML issue, so it should probably be dealt with there.

**Comment by xabbuh** [comment]
@fabpot The only thing is that technically, you can have a service id starting with `@`.

But anyway, I think it makes sense to restrict the list of allowed characters and deprecate everything else. So :+1: for doing this check for the YAML part only.

**Comment by fabpot** [comment]
Thank you @javiereguiluz.

**Pull request closed without merging**